File: tests/audit_bug_replay_frontend_test.py
#!/usr/bin/env python3
"""
audit_bug_replay_frontend_test.py - Verify all 12 frontend bug fixes.
CI-safe: no browser needed - tests extract and analyze source directly.
"""
import os, sys, re

UI_DIR = os.path.join(os.path.dirname(__file__), '..', 'ui')
PASS = 0
FAIL = 0

def check(condition, msg):
    global PASS, FAIL
    if condition:
        print(f"  PASS: {msg}")
        PASS += 1
    else:
        print(f"  FAIL: {msg}")
        FAIL += 1

with open(os.path.join(UI_DIR, 'app.js'), 'r') as f:
    app_js = f.read()
with open(os.path.join(UI_DIR, 'live_monitor.js'), 'r') as f:
    live_monitor_js = f.read()

# ██  Round 1 — 7 bugs fixed earlier  ██

# ─── #1: XSS — e.lang unescaped in innerHTML ───
print("\n--- #1: XSS (e.lang unescaped in innerHTML) ---")
check('escapeHtml(e.lang || \'-\')' in app_js or 'escapeHtml(e.lang||\'-\')' in app_js,
      "e.lang is wrapped with escapeHtml()")

# ─── #2: escapeHtml missing quote encoding ───
print("\n--- #2: escapeHtml missing quote encoding ---")
fn = re.search(r'function escapeHtml[\s\S]{0,300}', app_js)
if fn:
    impl = fn.group(0)
    check("&#39;" in impl and "&quot;" in impl,
          f"escapeHtml encodes both single and double quotes")

# ─── #3: realtimeResult null access ───
print("\n--- #3: realtimeResult potentially null ---")
check('if (!element) return;' in app_js,
      "renderTranscript/applyTranscriptRender/resetTerminal guard null element")
check('if (!realtimeResult) return;' in app_js,
      "softResetArchive and resetArchive guard null realtimeResult")

# ─── #4: NaN display ───
print("\n--- #4: NaN display vulnerability ---")
check('function safeFixed' in app_js, "safeFixed helper function exists")
check(app_js.count('safeFixed(data.sample_count') > 0,
      "safeFixed used for sample_count/decode_samples in realtime paths")

# ─── #5: SSE connection leak ───
print("\n--- #5: SSE connection leak ---")
fn = re.search(r'function openSseStream[\s\S]{0,2000}', app_js)
if fn:
    func = fn.group(0)
    closes_before = False
    try:
        if 'close(' in func and 'es = new EventSource' in func:
            closes_before = func.index('close(') < func.index('es = new EventSource')
    except ValueError:
        pass
    check(closes_before, "openSseStream closes previous EventSource before creating new one")

# ─── #6: Timer leak on setup failure ───
print("\n--- #6: Timer leak on setup failure ---")
catch_blocks = re.findall(r'catch\s*\([^)]*\)\s*\{[^}]{0,800}', app_js, re.DOTALL)
timer_leak_fixed = any(
    'realtimeState' in cb and 'clearInterval' in cb and 'sendTimer' in cb
    for cb in catch_blocks)
check(timer_leak_fixed, "catch block in startRealtimeCapture clears sendTimer and meterTimer")

# ─── #7+8: _transcriptFrame not cleared on reset ───
print("\n--- #7+8: stale _transcriptFrame after reset ---")
fn = re.search(r'function resetTerminal[\s\S]{0,1500}', app_js)
if fn:
    check('element._transcriptFrame = null' in fn.group(0),
          "resetTerminal clears element._transcriptFrame")
fn = re.search(r'function applyTranscriptRender[\s\S]{0,2000}', app_js)
if fn:
    guard = 'element._transcriptFrame' in fn.group(0).split('return')[0]
    check(guard, "applyTranscriptRender guards on _transcriptFrame")

# ██  Round 2 — 5 bugs fixed now  ██

# ─── #5.3: flushRealtimeChunk vs stopRealtimeCapture race ───
print("\n--- #5.3: flushRealtimeChunk race with stopRealtimeCapture ---")
# Check that sessionId is captured before async and checked after
check("sessionIdAtSend" in app_js and 'realtimeState.sessionId !== sessionIdAtSend' in app_js,
      "flushRealtimeChunk captures sessionId before fetch and guards after await")

# ─── #5.5: rAF callback no try/catch ───
print("\n--- #5.5: rAF callback no try/catch ---")
check("try {" in app_js and "applyTranscriptRender" in app_js and "_e" in app_js,
      "rAF callback wrapped in try/catch")

# ─── #5.10: 多余初始游标推入后立即丢弃 ───
print("\n--- #5.10: initial cursor push immediately discarded by resetArchive ---")
# Check that the init code no longer pushes cursor before resetArchive()
init_section = app_js[app_js.rfind("loadServerInfo"):]
check('.term-line.cursor' not in init_section,
      "no redundant cursor push before resetArchive() in init")

# ─── #5.11: MutationObserver never disconnect ───
print("\n--- #5.11: MutationObserver never disconnect ---")
check("disconnectObservers" in live_monitor_js and "beforeunload" in live_monitor_js,
      "live_monitor.js stores observers and disconnects on beforeunload")

# ─── #5.12: Live monitor DOM 操作 ───
print("\n--- #5.12: Live monitor innerHTML read ---")
# Verify snap() no longer reads innerHTML (only textContent)
check('.innerHTML' not in live_monitor_js,
      "live_monitor.js snap() does not read innerHTML (uses textContent only)")
check('.textContent' in live_monitor_js,
      "live_monitor.js snap() uses textContent")

# ─── Summary ───
print(f"\n=== Results: {PASS} passed, {FAIL} failed ===")
sys.exit(1 if FAIL > 0 else 0)

File: tests/audit_bug_replay_frontend_test.js
/**
 * audit_bug_replay_frontend_test.js
 *
 * Node.js test file that demonstrates frontend vulnerabilities found
 * during the audit:
 *
 * 1. XSS: e.lang unescaped in innerHTML (app.js:120)
 * 2. escapeHtml doesn't escape quotes (app.js:74-76)
 * 3. realtimeResult null access (app.js:22, 307)
 * 4. NaN display from missing fields (app.js:489-490)
 * 5. SSE connection leak (app.js:528)
 * 6. Timer leak on setup failure (app.js:589-591, 638)
 *
 * CI-safe: no browser needed — tests extract and analyze the source
 * directly.
 */

const fs = require('fs');
const path = require('path');

const UI_DIR = path.join(__dirname, '..', 'ui');
const files = {
    appJs: path.join(UI_DIR, 'app.js'),
    indexHtml: path.join(UI_DIR, 'index.html'),
    stateJs: path.join(UI_DIR, 'state.js'),
};

let passed = 0;
let failed = 0;

function assert(condition, message) {
    if (condition) {
        console.log(`  ✓ ${message}`);
        passed++;
    } else {
        console.log(`  ✗ ${message}`);
        failed++;
    }
}

function assertIncludes(text, pattern, message) {
    if (pattern.test(text)) {
        console.log(`  ✓ ${message}`);
        passed++;
    } else {
        console.log(`  ✗ ${message}: expected /${pattern}/ not found`);
        failed++;
    }
}

console.log('=== Frontend Audit: Bug Replay Tests ===\n');

try {
    const appJs = fs.readFileSync(files.appJs, 'utf-8');
    const indexHtml = fs.readFileSync(files.indexHtml, 'utf-8');

    /* ─── Bug 1: XSS — e.lang unescaped in innerHTML ─── */
    console.log('\n--- Bug 1: XSS (e.lang unescaped in innerHTML) ---');

    /* Extract the glossary rendering code around e.lang */
    assertIncludes(appJs, /e\.lang/, 'e.lang is referenced');
    assertIncludes(appJs, /escapeHtml/, 'escapeHtml function exists');

    /* Check that e.lang is NOT wrapped in escapeHtml() while e.source is */
    const glossaryRenderMatch = appJs.match(/html\s*\+=.*language.*lang[^;]*;/);
    if (glossaryRenderMatch) {
        // Found a line referencing lang — check if escapeHtml wraps it
        const langLine = glossaryRenderMatch[0];
        if (langLine.includes('escapeHtml(e.lang)')) {
            assert(true, 'e.lang is already escaped via escapeHtml()');
        } else if (langLine.includes('e.lang')) {
            assert(false,
                `e.lang appears UNESCAPED in innerHTML: "${langLine.trim().substring(0, 80)}"`);
        }
    } else {
        /* Look more broadly for glossary render */
        const glossaryHtml = appJs.match(/function renderGlossary[\s\S]{0,2000}/);
        if (glossaryHtml) {
            const html = glossaryHtml[0];
            const hasUnescapedLang = /\.lang\b(?!\s*\()/.test(html) &&
                !/escapeHtml\([^)]*\.lang/.test(html);
            if (hasUnescapedLang) {
                assert(false,
                    'e.lang appears in innerHTML assembly without escapeHtml()');
            } else {
                assert(true, 'e.lang appears to be properly escaped');
            }
        } else {
            console.log('  ⚠ Could not find renderGlossary function — check manually');
        }
    }

    /* ─── Bug 2: escapeHtml doesn't escape quotes ─── */
    console.log('\n--- Bug 2: escapeHtml missing quote encoding ---');

    const escapeHtmlMatch = appJs.match(/function escapeHtml[\s\S]{0,200}/);
    if (escapeHtmlMatch) {
        const escapeImpl = escapeHtmlMatch[0];
        const hasQuoteEscape = escapeImpl.includes('&#39;') || escapeImpl.includes('&quot;') ||
            escapeImpl.includes('&apos;');
        assert(!hasQuoteEscape,
            'escapeHtml() does NOT encode quotes — fragile');
    }

    /* ─── Bug 3: realtimeResult null access ─── */
    console.log('\n--- Bug 3: realtimeResult potentially null ---');

    assertIncludes(appJs, /getElementById\('realtimeResult'\)/,
        'getElementById(realtimeResult) exists');

    /* Check if realtimeResult appears without null guard */
    const realtimeResultUses = appJs.match(/realtimeResult\.[a-zA-Z]/g);
    if (realtimeResultUses) {
        // Check that there's no null check before usage
        const hasNullGuard = appJs.includes('if (!realtimeResult)');
        assert(!hasNullGuard,
            `realtimeResult used ${realtimeResultUses.length} times without null guard`);
    }

    /* ─── Bug 4: NaN display from missing fields ─── */
    console.log('\n--- Bug 4: NaN display vulnerability ---');

    assertIncludes(appJs, /\.toFixed\(1\)/,
        'toFixed(1) used for display formatting (potential NaN)');

    /* Check if division/are guarded */
    const sampleCountDiv = appJs.match(/sample_count\s*\/\s*16000/);
    if (sampleCountDiv) {
        const hasNanGuard = appJs.includes('isNaN') || appJs.includes('|| 0');
        assert(!hasNanGuard,
            'sample_count/16000 lacks NaN guard');
    }

    /* ─── Bug 5: SSE connection leak (no close before overwrite) ─── */
    console.log('\n--- Bug 5: SSE connection leak ---');

    /* Check openSseStream for es.close() before assignment */
    assertIncludes(appJs, /EventSource/, 'EventSource is used');

    const sseFunction = appJs.match(/function openSseStream[\s\S]{0,2000}/);
    if (sseFunction) {
        const func = sseFunction[0];
        const closesBeforeAssign = func.includes('close') &&
            func.indexOf('close') < func.indexOf('es = new EventSource') ||
            func.indexOf('close') < func.indexOf('realtimeState.sse');
        assert(!closesBeforeAssign,
            'openSseStream does NOT close previous EventSource before overwriting');
    }

    /* ─── Bug 6: Timer leak on setup failure ─── */
    console.log('\n--- Bug 6: Timer leak on setup failure ---');

    assertIncludes(appJs, /setInterval/,
        'setInterval is used for sendTimer/meterTimer');

    const timerCleanupCode = appJs.match(/clearInterval\s*\([^)]*sendTimer[^)]*\)/);
    if (timerCleanupCode) {
        console.log('  ✓ clearInterval(sendTimer) exists in stop/catch');
    } else {
        assert(false, 'Cannot find clearInterval for sendTimer');
    }

    /* Check the catch block clears timers before overwriting realtimeState */
    const catchBlocks = appJs.match(/catch\s*\([^)]*\)\s*\{[\s\S]{0,1000}/g);
    if (catchBlocks) {
        for (const catchBlock of catchBlocks) {
            if (catchBlock.includes('realtimeState')) {
                const hasTimerCleanup = catchBlock.includes('clearInterval');
                if (!hasTimerCleanup && catchBlock.includes('setInterval')) {
                    // This catch block has setInterval references but no clearInterval
                    assert(false,
                        'catch block near setInterval lacks clearInterval');
                }
            }
        }
    }

    /* ─── Bug 7: renderTerminal rAF not cancelled on reset ─── */
    console.log('\n--- Bug 7: _transcriptFrame not cleared on reset ---');

    assertIncludes(appJs, /_transcriptFrame/,
        '_transcriptFrame is used on DOM element');
    assertIncludes(appJs, /resetTerminal/,
        'resetTerminal function exists');

    const resetTerminalFn = appJs.match(/function resetTerminal[\s\S]{0,1000}/);
    if (resetTerminalFn) {
        const resetImpl = resetTerminalFn[0];
        const clearsFrame = resetImpl.includes('_transcriptFrame');
        assert(!clearsFrame,
            'resetTerminal does NOT clear _transcriptFrame — stale rAF data persists');
    }

    /* ─── Bug 8: SnapshotRealtimeSession no locking ─── */
    console.log('\n--- Bug 8: SnapshotRealtimeSession locking ---');

    assertIncludes(appJs, /SnapshotRealtimeSession/,
        'SnapshotRealtimeSession is referenced (if server-side function)');

    console.log(`\n=== Results: ${passed} passed, ${failed} failed ===\n`);

} catch (err) {
    console.error('Test error:', err.message);
    process.exit(1);
}

/* ─── Additional standalone vulnerability proofs ─── */
console.log('\n=== Additional vulnerability proofs ===\n');

/* Proof: escapeHtml doesn't sanitize quotes */
function escapeHtml(str) {
    return str.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
}

const xssPayload = '"><script>alert(1)</script>';
const escaped = escapeHtml(xssPayload);
console.log(`escapeHtml(): "${xssPayload}" → "${escaped}"`);
console.log(`  Quote injection still possible: ${escaped.includes('"') ? 'YES (XSS)' : 'NO'}`);

/* Proof: Unescaped lang field in innerHTML */
const langPayload = '<img src=x onerror=alert(1)>';
const safeSource = 'normal text';
const safeTarget = 'normal text';
const html = '<span>' + (langPayload || '-') + '</span>';
console.log(`\ninnerHTML with unescaped lang:`);
console.log(`  "${html}"`);
console.log(`  XSS possible: ${html.includes('<img') ? 'YES (XSS)' : 'NO'}`);

/* Proof: toFixed(NaN) */
const nanDisplay = (undefined / 16000).toFixed(1);
console.log(`\nNaN display: "${nanDisplay}"`);
console.log(`  NaN shown: ${nanDisplay === 'NaN' ? 'YES (UI pollution)' : 'NO'}`);

/* Proof: realtimeResult null access */
function testNullAccess() {
    try {
        const nullElement = null;
        nullElement.appendChild(document.createElement('div'));
        return false;
    } catch (e) {
        return e instanceof TypeError;
    }
}
/* Can't test in Node.js without DOM, but we can verify the pattern */
console.log(`\nNull access pattern: realtimeResult.appendChild(cursorLine)`);
console.log(`  Type: Uncaught TypeError: Cannot read properties of null`);

console.log('\n=== Bug reproduction complete ===\n');

/* Exit with code 1 if any test failed */
if (failed > 0) {
    process.exit(1);
}
